chunk/aws_storage_client.go

```diff
@@ -222,7 +273,95 @@ func (d dynamoClientAdapter) QueryPages(ctx context.Context, entry IndexEntry, c
 	return nil
 }
 
-func (d dynamoClientAdapter) ListTables() ([]string, error) {
+func (a awsStorageClient) GetChunk(ctx context.Context, key string) ([]byte, error) {
+	var resp *s3.GetObjectOutput
+	err := instrument.TimeRequestHistogram(ctx, "S3.GetObject", s3RequestDuration, func(_ context.Context) error {
+		var err error
+		resp, err = a.S3.GetObject(&s3.GetObjectInput{
+			Bucket: a.bucketName,
+			Key:    aws.String(key),
+		})
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	buf, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+	return instrument.TimeRequestHistogram(ctx, "S3.PutObject", s3RequestDuration, func(_ context.Context) error {
+		_, err := a.S3.PutObject(&s3.PutObjectInput{
+			Body:   bytes.NewReader(buf),
+			Bucket: a.bucketName,
+			Key:    aws.String(key),
+		})
+		return err
+	})
+}
+
+type dynamoDBWriteBatch map[string][]*dynamodb.WriteRequest
+
+func (b dynamoDBWriteBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	item := map[string]*dynamodb.AttributeValue{
+		hashKey:  {S: aws.String(hashValue)},
+		rangeKey: {B: rangeValue},
+	}
+
+	if value != nil {
+		item[valueKey] = &dynamodb.AttributeValue{B: value}
+	}
+
+	b[tableName] = append(b[tableName], &dynamodb.WriteRequest{
+		PutRequest: &dynamodb.PutRequest{
+			Item: item,
+		},
+	})
+}
+
+type dynamoDBReadBatch []map[string]*dynamodb.AttributeValue
+
+func (b dynamoDBReadBatch) Len() int {
+	return len(b)
+}
+
+func (b dynamoDBReadBatch) RangeValue(i int) []byte {
+	return b[i][rangeKey].B
+}
+
+func (b dynamoDBReadBatch) Value(i int) []byte {
+	chunkValue, ok := b[i][valueKey]
+	if !ok {
+		return nil
+	}
+	return chunkValue.B
+}
+
+type dynamoTableClient struct {
```

Comment:
could we embed this in `awsStorageClient` to reduce some of the duplication between `newDynamoDBConfig` and `NewAWSStorageClient`?